src/test/ui/issues/issue-66851.rs

```diff
@@ -0,0 +1,17 @@
+// run-pass
+// compile-flags: -Zmir-opt-level=2
+
```

Comment:
```suggestion

// This used to mis-compile because the mir-opt `SimplifyArmIdentity`
// did not check that the types matched up in the `Ok(r)` branch.

```